### DISCUSSION TITLE
Get previous record state to compare it with update via SSE.

### DISCUSSION THREAD

**Discussion by sotnii** [pocketbase/pocketbase#2732] [Q&A] [answered]
I'm wondering how can I get only changes, or maybe previous state of the record to compare them. 

I have this kind of field that can be appended with other table's ids, and I need to know which fields were added, and which were already present before the update. 
```json
{
  "source": [ "fneh6c7n95q537g", "pdwnzoa5sf88mf8" ]
}
```

Any suggestions how to do this? Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
Currently we don't sent the previous record state with the SSE event mainly because:

- not everyone need it and sending it by default may increase the messages latency since record models can be of any size
- the client may not have access to the previous/new record state

If you want to compare prev and new record state, you'll have to keep track of the record client-side (eg. fetch the original record before starting realtime subscription).

Alternatively, if you [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/), you can attach to the [`OnModelAfterUpdate` hook](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks) and access the previous record state using [`Record.OriginalCopy()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.5/models#Record.OriginalCopy), something like:

```js
app.OnModelAfterUpdate("yourCollection").Add(func(e *core.ModelEvent) error {
	record := e.Model.(*models.Record)

	oldRecord := record.OriginalCopy()

	oldName := oldRecord.GetString("name")
	
	log.Println(oldName)

	return nil
})
```

---

_With the next release (_there are no ETAs yet_) you'll be able to do the above also with JavaScript, which will look something like:_

```js
$app.onModelAfterUpdate("yourCollection").add((e) => {
	const oldRecord = e.model.originalCopy()
	
	const oldName = oldRecord.get("name")
	
	console.log(oldName)
})
```

**Reply by sotnii** [reply]
I think it would be nice to add some option when subscribing to realtime events like `withOriginalCopy`. But anyways, thanks for the reply, I'm it'll help!

**Reply by sotnii** [reply]
Is there a way to eastablish custom SSE connection so I could send updates with original copy myself, maybe?

**Reply by ganigeorgiev** [reply]
There is no safe option to request for a original copy from client-side. We can't add support for `withOriginalCopy` because the second listed issue remains, aka. the request client may not have had been allowed to access the prev model state (_for example a "post" record is updated from "draft" to "public"_).

**Reply by sotnii** [reply]
Ok, I see

**Reply by benallfree** [reply]
@ganigeorgiev FYI it appears that in the JS hooks version, ModelEvent passes a Model rather than a Record https://pocketbase.io/jsvm/interfaces/core.ModelEvent.html#model so there is no originalCopy - at least according to docs and TS typings. `e.model.originalCopy()` does return the previous model though.

**Reply by ganigeorgiev** [reply]
@benallfree This is working as intended although I agree that it could be confusing. The `onModel*` hooks fire for every `$app.dao` operation with struct that implements the `models.Model` interface (collections, admins, params, records, etc.). In Go usually you have to type assert it to `*models.Record` to access the specific record fields and methods but in the goja runtime you can access directly the fields behind the interface.

**Reply by benallfree** [reply]
@ganigeorgiev Are there `onModel*` cases where `originalCopy` is not available?

**Reply by ganigeorgiev** [reply]
Yes, on any other non-`Record` model operation, eg. `Collection` schema change or app settings update.
If you specify a hook "tag" (aka. the collection name(s) you are targeting) as last argument it is expected that `e.model` will be always `Record`.

**Comment by sotnii** [comment]
Just in case if anyone bumps sinto the same issue, here's my code to send custom events:
```go
app.OnRealtimeAfterSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
	msg := subscriptions.Message{
		// Just getting the last subscription
		Name: e.Subscriptions[len(e.Subscriptions)-1],
		Data: "{ \"action\": \"my_custom_event\" }",
	}

	e.Client.Channel() <- msg
	return nil
})
```

Then you can integrate it with `app.OnModelAfterUpdate`and send events to clients with custom data attached.
